From: Kirill Rd <theoden8@tutamail.com>
Subject: [PATCH] http: allow passing in pre-existing connection bev (#1722)

This patch adds functionality to pass a pre-existing connection
as a bufferevent to `evhttp_connection_base_bufferevent_reuse_new`.

When the bufferevent has an existing fd, the evcon starts
in state `EVCON_IDLE` so that requests can be made immediately.

Pantavisor uses it to run Hub requests over a TLS bufferevent filtered on
top of an HTTP proxy CONNECT tunnel. Backported to 2.2.1-alpha without the
regress test changes.

Upstream-Status: Backport [https://github.com/libevent/libevent/commit/98b7ca30b2daec33fb98e7d16ee8891b36d32f50]
Signed-off-by: Kirill Rodriguez <theoden8@tutamail.com>
Co-authored-by: Wladimir J. van der Laan <laanwj@gmail.com>
Co-authored-by: Azat Khuzhin <a3at.mail@gmail.com>
---
diff --git a/http.c b/http.c
index 98fcfbc9..0f971521 100644
--- a/http.c
+++ b/http.c
@@ -2534,6 +2534,18 @@ evhttp_connection_new(const char *address, ev_uint16_t port)
 	return (evhttp_connection_base_new(NULL, NULL, address, port));
 }
 
+/* We were passed a bev with file descriptor set.
+ * Assume that this is an already-open connection that we
+ * can start sending requests on.
+ */
+static int
+evhttp_connection_set_existing_(struct evhttp_connection *evcon, struct bufferevent* bev)
+{
+	evcon->state = EVCON_IDLE;
+	evcon->flags |= EVHTTP_CON_OUTGOING;
+	return 0;
+}
+
 static struct evhttp_connection *
 evhttp_connection_new_(struct event_base *base, struct bufferevent* bev)
 {
@@ -2588,6 +2600,35 @@ evhttp_connection_new_(struct event_base *base, struct bufferevent* bev)
 	return (NULL);
 }
 
+struct evhttp_connection *
+evhttp_connection_base_bufferevent_reuse_new(struct event_base *base, struct evdns_base *dnsbase, struct bufferevent* bev)
+{
+	struct evhttp_connection *evcon = NULL;
+	if (bev == NULL)
+		goto error;
+
+	evcon = evhttp_connection_new_(base, bev);
+
+	if (evcon == NULL)
+		goto error;
+
+	if (evhttp_connection_set_existing_(evcon, bev))
+		goto error;
+
+	evcon->dns_base = dnsbase;
+	evcon->address = NULL;
+	evcon->port = 0;
+#ifndef _WIN32
+	evcon->unixsocket = NULL;
+#endif
+
+	return (evcon);
+ error:
+	if (evcon != NULL)
+		evhttp_connection_free(evcon);
+	return (NULL);
+}
+
 #ifndef _WIN32
 struct evhttp_connection *
 evhttp_connection_base_bufferevent_unix_new(struct event_base *base, struct bufferevent* bev, const char *unixsocket)
diff --git a/include/event2/http.h b/include/event2/http.h
index d1790613..4d3033ae 100644
--- a/include/event2/http.h
+++ b/include/event2/http.h
@@ -699,6 +699,24 @@ EVENT2_EXPORT_SYMBOL
 struct evhttp_connection *evhttp_connection_base_bufferevent_unix_new(
 	struct event_base *base, struct bufferevent* bev, const char *path);
 
+/**
+ * Create and return a connection object that can be used to for making HTTP
+ * requests. The connection attempts to reuse an existing connection that was
+ * already established with bufferevent.
+ *
+ * @param base the event_base to use for handling the connection
+ * @param dnsbase the dns_base to use for resolving host names; if not
+ *     specified host name resolution will block.
+ * @param bev a bufferevent to use for connecting to the server. A fd
+ *     is already set on the bufferevent, it will be assumed that this
+ *     connection is already open and ready to send requests.
+ * @return an evhttp_connection object that can be used for making requests or
+ *   NULL on error
+ */
+EVENT2_EXPORT_SYMBOL
+struct evhttp_connection *
+evhttp_connection_base_bufferevent_reuse_new(struct event_base *base, struct evdns_base *dnsbase, struct bufferevent* bev);
+
 /**
  * Return the bufferevent that an evhttp_connection is using.
  */
